doc/BACKUP.md

```diff
@@ -0,0 +1,391 @@
+`$LIGHTNINGDIR` together with the `hsm_secret`.
+
+You can also use any automated backup system as long as it includes the
+`lightningd.sqlite3` file (and optionally `hsm_secret`, but note that
+as a secret key, thieves getting a copy of your backups may allow them
+to steal your funds, even in-channel funds).
+
+This backup method is undesirable, since it cannot recover the following
+channels:
+
+* Channels with peers that do not support `option_dataloss_protect`.
+  * Most nodes on the network already support `option_dataloss_protect`
+    as of November 2020.
+  * If the peer does not support this option, then the entire channel
+    funds will be revoked by the peer.
+  * Specially-coded peers can pretend to support this option, but end
+    up taking the entire channel funds (or otherwise more than what it
+    should have) when you try to recover.
+    Standard releases of Lightning Network node software do not do this
+    (they just support this option according to the standard) but note
+    that anyone can modify their node software, especially with our
+    well-documented open-source code.
+* Channels created *after* the copy was made are not recoverable.
+  * Data for those channels does not exist in the backup, so your node
+    cannot recover them.
+
+Because of the above, this strategy is discouraged: you *can* potentially
+lose all funds in open channels.
+
+However, again, note that a "no backups #reckless" strategy leads to
+*definite* loss of funds, so you should still prefer *this* strategy rather
+than having *no* backups at all.
+
+Even if you have the better options below, you might still want to do
+this as a worst-case fallback, as long as you:
+
+* Attempt to recover using the other backup options below first.
+  Any one of them will be better than this backup option.
+* Recover by this method **ONLY** as a ***last*** resort.
+* Recover using the most recent backup you can find.
+  Take time to look for the most recent available backup.
+
+Again, this strategy can lead to only ***partial*** recovery of funds,
+or even to complete failure to recover, so use the other methods below
+first to recover!
+
+### Backing Up While `lightningd` Is Running
+
+Since `sqlite3` will be writing to the file while `lightningd` is running,
+`cp`ing the `lightningd.sqlite3` file while `lightningd` is running may
+result in the file not being copied properly if `sqlite3` happens to be
+committing database transactions at that time, potentially leading to a
+corrupted backup file that cannot be recovered from.
+
+Instead, the **proper** way to do this would be:
+
+* `echo "VACUUM INTO 'backup.sqlite3';" | sqlite3 lightningd.sqlite3`
+* `sync`
+* `mv backup.sqlite3 ${YOURBACKUPLOCATION}/lightningd.sqlite3`
+
+This creates a consistent snapshot of the database, sampled in a
+transaction, that is assured to be openable later by `sqlite3`.
+The operation of the `lightningd` process will be paused while `sqlite3`
```

Comment:
***THIS IS FACTUALLY INCORRECT***.

I did a stress test where a program continuously makes many small transactions to update a database.  When you do something like `.backup 'backup.sqlite3'` or `VACUUM INTO `backup.sqlite3';` in a *separate* `sqlite3`, then either the main program gets an `SQLITE_BUSY`, ***or*** the separate `sqlite3` returns a "Database is locked" error.

I suspect that an `SQLITE_BUSY` will cause `lightningd` to crash, so this is probably strongly not recommended to use a separate process to back up.

We can do:

* Expose an `sqlite3_backup` command which performs the `VACUUM INTO` query inside `lightningd`.
* Call into `sqlite3_busy_timeout` with a "reasonable" large value, say 5000 (5 seconds). (Equivalently, call `PRAGMA busy_timeout = 5000;` query at the same time we do `PRAGMA foreign_keys=on;`).  Then, even if the `.backup` or `VACUUM INTO` takes up to 5 seconds to back up, `lightningd` will just wait.

The latter solution is easier but there is always the possibility that the backup process will take more than whatever timeout we select.  The earlier solution is more reliable and there is no timeout, but note that if say the target location is slow (e.g. an NFS mount) then `lightningd` will suspend indefinitely.

Thoughts?

Reply:
Thank you @ZmnSCPxj for ideas and investigating possible issues with running backup in separate `sqlite3` process!

Exposing the `sqlite3_backup` command would be great! It sounds like a very nice and clear way to do backup.